osquery/tables/smart/smart_drives.cpp

```diff
@@ -0,0 +1,219 @@
+  std::vector<std::string> results;
+
+  auto devices = SQL::selectAllFrom("pci_devices");
+  for (const auto& device : devices) {
+    if (device.find("pci_class") != device.end() &&
+        device.at("pci_class") == "Mass storage controller" &&
+        device.find("driver") != device.end()) {
+      auto i =
+          std::lower_bound(results.begin(), results.end(), device.at("driver"));
+      if (i == results.end() || device.at("driver") < *i) {
+        results.insert(i, std::move(device.at("driver")));
+      }
+    }
+  }
+
+  return results;
+}
+
+/// Gets supported hardwareDriver from system storage controllers.
+static inline void getSmartCtlDeviceType(
+    const std::vector<std::string>& storageDrivers,
+    std::vector<hardwareDriver>& types) {
+  for (auto const& driver : storageDrivers) {
+    if (kSMARTExplicitDriverToDevice.find(driver) !=
+        kSMARTExplicitDriverToDevice.end()) {
+      types.push_back(kSMARTExplicitDriverToDevice.at(driver));
+    }
+  }
+}
+
+/// Utility function for traversing system devices.
+void walkDevices(std::function<bool(libsmartctl::Client& c,
+                                    const std::string& devname,
+                                    hardwareDriver* type)> handleDevF) {
+  if (getuid() || geteuid()) {
+    LOG(WARNING) << "Need root access for smart information";
+  }
+
+  QueryData results;
+  libsmartctl::Client c;
+
+  auto storageDrivers = getStorageCtlerClassDrivers();
+
+  std::vector<hardwareDriver> types;
+  getSmartCtlDeviceType(storageDrivers, types);
+
+  if (types.size() > 1) {
+    LOG(WARNING) << "Found more than 1 hardware storage controller:" << types
+                 << "; only handling the first";
+    return;
+  }
+
+  hardwareDriver* type{nullptr};
+  if (types.size() == 1) {
+    type = &(types[0]);
+  }
+
+  auto blkDevices = SQL::selectAllFrom("block_devices");
+  for (const auto& device : blkDevices) {
+    if (device.find("size") != device.end() && device.at("size") != "0" &&
+        device.at("size") != "" && device.find("name") != device.end()) {
+      if (handleDevF(c, device.at("name"), type)) {
+        break;
+      };
+    }
+  }
+}
+
+QueryData genSmartInfo(QueryContext& context) {
+  QueryData results;
+  /* hwInfo is for tracking info retrieve with an explicit HW controller.  It is
+   * indexed by serial_number, since that's how you correlate the data with
+   * auto-detect retrieved SMART info. */
+  std::map<std::string, Row> hwInfo;
+
+  bool found{false};
+  walkDevices([&](libsmartctl::Client& c,
+                  const std::string& devname,
+                  hardwareDriver* type) {
+    // Get auto info..
+    auto resp = c.getDevInfo(devname, "");
+    if (resp.err != NOERR) {
+      LOG(INFO) << "There was an error retrieving drive information: "
+                << libsmartctl::errStr(resp.err);
+
+    } else {
+      resp.content["device_name"] = devname;
+      results.push_back(std::move(resp.content));
```

Comment:
I'm not 100% I follow here, you do a `push_back`, but down below you're still modifying `resp` contents and don't do anything with it later on before the scope ends, is this intended? Seems like this `push_back` should happen later?